glnx-errors.h

```diff
@@ -25,25 +25,101 @@
 
 G_BEGIN_DECLS
 
+/* Set @error with G_IO_ERROR/G_IO_ERROR_FAILED.
+ *
+ * This function returns %FALSE so it can be used conveniently in a single
+ * statement:
+ *
+ * ``
+ *   if (strcmp (foo, "somevalue") != 0)
+ *     return glnx_throw (error, "key must be somevalue, not '%s'", foo);
+ * ```
+ */
+static inline gboolean G_GNUC_PRINTF (2,3)
+glnx_throw (GError **error, const char *fmt, ...)
+{
+  if (error == NULL)
+    return FALSE;
+
+  va_list args;
+  va_start (args, fmt);
+  GError *new = g_error_new_valist (G_IO_ERROR, G_IO_ERROR_FAILED, fmt, args);
+  va_end (args);
+  g_assert_no_error (*error);
```

Comment:
For something like this I think it deserves careful review!  That's a good point.  I originally did the assertion because I only was looking at `g_set_error()` as handling the overwritten warning, but you're definitely right, propagate should work.  Will do.